satellite/gc/service.go

```diff
@@ -0,0 +1,166 @@
+	// value for InitialPieces currently based on average pieces per node
+	InitialPieces     int64   `help:"the initial number of pieces expected for a storage node to have, used for creating a filter" releaseDefault:"400000" devDefault:"10"`
+	FalsePositiveRate float64 `help:"the false positive rate used for creating a filter" releaseDefault:"0.1" devDefault:"0.1"`
+}
+
+// Service implements the garbage collection service
+type Service struct {
+	log             *zap.Logger
+	config          Config
+	transport       transport.Client
+	overlay         overlay.DB
+	lastSendTime    time.Time
+	lastPieceCounts atomic.Value
+}
+
+// NewService creates a new instance of the gc service
+func NewService(log *zap.Logger, config Config, transport transport.Client, overlay overlay.DB) *Service {
+	var lastPieceCounts atomic.Value
+	lastPieceCounts.Store(map[storj.NodeID]int{})
+
+	return &Service{
+		log:             log,
+		config:          config,
+		transport:       transport,
+		overlay:         overlay,
+		lastPieceCounts: lastPieceCounts,
+	}
+}
+
+// NewPieceTracker instantiates a piece tracker
+func (service *Service) NewPieceTracker() *PieceTracker {
+	// Creation date of the gc bloom filter - the storage nodes shouldn't delete any piece newer than this.
+	filterCreationDate := time.Now().UTC()
+
+	if !service.config.Enabled || filterCreationDate.Before(service.lastSendTime.Add(service.config.Interval)) {
+		return nil
+	}
+
+	return &PieceTracker{
+		log:                service.log.Named("piecetracker"),
+		filterCreationDate: filterCreationDate,
+		initialPieces:      service.config.InitialPieces,
+		falsePositiveRate:  service.config.FalsePositiveRate,
+		retainInfos:        make(map[storj.NodeID]*RetainInfo),
+		pieceCounts:        service.lastPieceCountsValue(),
+		overlay:            service.overlay,
+	}
+}
+
+// Send sends the piece retain requests to all storage nodes
+func (service *Service) Send(ctx context.Context, pieceTracker *PieceTracker, cb func()) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	service.lastSendTime = time.Now().UTC()
+
+	go func() {
+		piecesCounts, err := service.sendRetainRequests(ctx, pieceTracker)
+		if err != nil {
+			service.log.Error("error sending retain infos", zap.Error(err))
+		}
+
+		service.lastPieceCounts.Store(piecesCounts)
+		cb()
+	}()
+
+	return nil
+}
+
+func (service *Service) sendRetainRequests(
+	ctx context.Context, pieceTracker *PieceTracker,
+) (pieceCounts map[storj.NodeID]int, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	pieceCounts = make(map[storj.NodeID]int, service.lastPieceCountsNumNodes())
+
+	var errList errs.Group
+	for id, retainInfo := range pieceTracker.GetRetainInfos() {
+		err := service.sendOneRetainRequest(ctx, id, retainInfo, pieceCounts)
+		if err != nil {
+			errList.Add(err)
+		}
+	}
+	return pieceCounts, errList.Err()
+}
+
+func (service *Service) sendOneRetainRequest(
```

Comment:
Why is the `pieceCounts` passed in here?

Reply:
Because it's the way to update map without having mutex, as it's created by its caller https://github.com/storj/storj/pull/2366/files/2d6ce291e8cb8bf8ddc790c14718455539e2a149#diff-f1ff4261b7834d51a195dcfed9762328R104

Reply:
Why does it have to be updated inside of the sendone in the first place?

Reply:
yes, right, it could be in the caller.

I didn't realize when I did the refactoring.

Reply:
Why does it have to be updated in this method?

Reply:
As I said, the caller can do the update, that was my mistake when I refactored the code.

Reply:
Sorry, I missed the second comment, github didn't auto-refresh for some reason.